### PR TITLE
fall back to full-graph aggregation variant for MFGs with large fanout values

### PR DESCRIPTION
**Pull request by tingyu66**
This PR allows cugraph-ops models to accept dgl blocks with large in-degrees. 

There will be some performance penalty due to creating an extra buffer for the augmented colptr.

CC: @VibhuJawa

### CODE REVIEW

File: python/cugraph-dgl/cugraph_dgl/nn/conv/gatconv.py [hunk 1]
```diff
@@ -149,9 +150,19 @@ def forward(
...
-            _graph = ops.make_mfg_csr(
-                g.dstnodes(), offsets, indices, max_in_degree, g.num_src_nodes()
-            )
+
+            if max_in_degree < self.MAX_IN_DEGREE_MFG:
+                _graph = ops.make_mfg_csr(
+                    g.dstnodes(), offsets, indices, max_in_degree, g.num_src_nodes()
+                )
+            else:
+                offsets_fg = torch.empty(
```
Comment (stadlmax) [inline python/cugraph-dgl/cugraph_dgl/nn/conv/gatconv.py:159]: where is there a difference between the offsets?

Comment (tingyu66) [reply python/cugraph-dgl/cugraph_dgl/nn/conv/gatconv.py:159]: offsets_mfg has a shape of (n_out_nodes+1, ), while offsets_fg has a shape of (n_in_nodes+1).
Hence, we have to pad (n_in_nodes - n_out_nodes) elements at the end of offset to form offset_fg.

Comment (stadlmax) [reply python/cugraph-dgl/cugraph_dgl/nn/conv/gatconv.py:159]: ah yes, sure ..., the kernels technically would work with offsets to be `n_out_nodes + 1`, however, as we use the offsets currently to determine `n_nodes` in our bindings we need it for that, my bad

Comment (stadlmax) [reply python/cugraph-dgl/cugraph_dgl/nn/conv/gatconv.py:159]: Either way: for the current GraphCast involvement, we would need proper support for any kind of bipartite-graphs (full graphs are just a special case of these when it comes to message passing) so we might address this rather quickly

Comment (tingyu66) [reply python/cugraph-dgl/cugraph_dgl/nn/conv/gatconv.py:159]: > ah yes, sure ..., the kernels technically would work with offsets to be `n_out_nodes + 1`, however, as we use the offsets currently to determine `n_nodes` in our bindings we need it for that, my bad

Oh, I though this restriction comes all the way from the C++ APIs, as noted here:
https://github.com/rapidsai/cugraph-ops/blob/ddd07dded5ddfdc88f22121b97adb9cc3ca4dbbd/cpp/include/cugraph-ops/graph/format.hpp#L82-L87

Regardless, support for generic bipartite graphs would be extremely beneficial to upstream integration.

Comment (stadlmax) [reply python/cugraph-dgl/cugraph_dgl/nn/conv/gatconv.py:159]: It is formulated like that in the comment. However, in the kernels themselves, you often just iterate over the offests (i.e. the out-nodes). Buffers from in-nodes are usually indexed like `input[indices[neigh]]` where `neigh` is between `offsets[current_out_node]` and `offsets[current_out_node + 1]`
In our CSR representation, offsets inherently refer to the out-nodes and indices to the in-nodes.

**Comment by VibhuJawa** [comment]
Can confirm that this **PR** resolves the large fan out based sampling cugraph was doing.  Also i am seeing a approx 34% boost in Perf on the pure modeling front.   Thanks for working on this @tingyu66 .  

cugraph-ops+DGL:
```python3
Running Epoch = 0 on rank = None
Epoch: 00000 | Loss: 48227.28 Total time 8.40 s | Model time 2.89 s
Running Epoch = 1 on rank = None
Epoch: 00001 | Loss:  453.38 Total time 7.00 s | Model time 1.82 s
Running Epoch = 2 on rank = None
Epoch: 00002 | Loss:  289.01 Total time 7.09 s | Model time 2.09 s
Running Epoch = 3 on rank = None
Epoch: 00003 | Loss:  209.70 Total time 6.94 s | Model time 1.81 s
Running Epoch = 4 on rank = None
Epoch: 00004 | Loss:  158.72 Total time 7.11 s | Model time 2.10 s
Running Epoch = 5 on rank = None
Epoch: 00005 | Loss:  128.66 Total time 7.33 s | Model time 2.09 s
Running Epoch = 6 on rank = None
Epoch: 00006 | Loss:  110.06 Total time 6.73 s | Model time 2.08 s
Running Epoch = 7 on rank = None
Epoch: 00007 | Loss:   95.46 Total time 7.49 s | Model time 2.37 s
Running Epoch = 8 on rank = None
Epoch: 00008 | Loss:   83.22 Total time 6.96 s | Model time 1.78 s
Running Epoch = 9 on rank = None
Epoch: 00009 | Loss:   74.69 Total time 7.17 s | Model time 2.13 s
```


DGL:
```python3
Running Epoch = 0 on rank = None
Epoch: 00000 | Loss: 1924890.39 Total time 10.19 s | Model time 4.28 s
Running Epoch = 1 on rank = None
Epoch: 00001 | Loss: 53358.80 Total time 8.28 s | Model time 3.21 s
Running Epoch = 2 on rank = None
Epoch: 00002 | Loss: 3267.85 Total time 8.38 s | Model time 3.22 s
Running Epoch = 3 on rank = None
Epoch: 00003 | Loss: 1215.29 Total time 8.40 s | Model time 3.17 s
Running Epoch = 4 on rank = None
Epoch: 00004 | Loss:  690.16 Total time 8.51 s | Model time 3.18 s
Running Epoch = 5 on rank = None
Epoch: 00005 | Loss:  464.87 Total time 8.58 s | Model time 3.16 s
Running Epoch = 6 on rank = None
Epoch: 00006 | Loss:  330.41 Total time 8.44 s | Model time 3.18 s
Running Epoch = 7 on rank = None
Epoch: 00007 | Loss:  245.18 Total time 8.32 s | Model time 3.11 s
Running Epoch = 8 on rank = None
Epoch: 00008 | Loss:  191.57 Total time 8.42 s | Model time 3.15 s
Running Epoch = 9 on rank = None
Epoch: 00009 | Loss:  151.37 Total time 8.53 s | Model time 3.19 s
```

**Review by VibhuJawa — APPROVED** [review]
LGTM

**Comment by VibhuJawa** [comment]
@tingyu66 , You might want to merge in `branch-23.04` and then add labels to make sure we pass CI

**Comment by tingyu66** [comment]
Superseded by #3360.

**Pull request closed without merging**